File: java/client/src/test/java/org/signal/libsignal/messagebackup/BackupForwardSecrecyTokenTest.kt
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.messagebackup

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.signal.libsignal.zkgroup.InvalidInputException

class BackupForwardSecrecyTokenTest {
  @Test
  fun testValidTokenCreation() {
    val validBytes = ByteArray(32) { 0x42 }
    val token = BackupForwardSecrecyToken(validBytes)
    assertNotNull(token)
    assertEquals(32, token.serialize().size)

    val retrievedBytes = token.serialize()
    assertArrayEquals(validBytes, retrievedBytes)
  }

  @Test(expected = InvalidInputException::class)
  fun testInvalidTokenCreationTooShort() {
    val invalidBytes = ByteArray(31) { 0x42 }
    BackupForwardSecrecyToken(invalidBytes)
  }

  @Test(expected = InvalidInputException::class)
  fun testInvalidTokenCreationTooLong() {
    val invalidBytes = ByteArray(33) { 0x42 }
    BackupForwardSecrecyToken(invalidBytes)
  }

  @Test
  fun testRoundTripSerialization() {
    // Use different hardcoded pattern to ensure we're not just getting lucky
    val originalBytes = ByteArray(32) { index -> (index % 256).toByte() }
    val token = BackupForwardSecrecyToken(originalBytes)

    val serialized = token.serialize()
    val reconstructedToken = BackupForwardSecrecyToken(serialized)

    assertNotNull(reconstructedToken)
    assertEquals(32, reconstructedToken.serialize().size)

    val reconstructedBytes = reconstructedToken.serialize()
    assertArrayEquals(originalBytes, reconstructedBytes)
  }
}

File: node/ts/test/BackupForwardSecrecyTokenTest.ts
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { assert, expect } from 'chai';
import { BackupForwardSecrecyToken } from '../AccountKeys.js';

describe('BackupForwardSecrecyToken', () => {
  it('can create a valid token and retrieve same bytes', () => {
    const validBytes = new Uint8Array(32).fill(0x42);
    const token = new BackupForwardSecrecyToken(validBytes);

    assert.exists(token);
    assert.equal(token.serialize().length, 32);

    const retrievedBytes = token.serialize();
    expect(retrievedBytes).to.deep.equal(validBytes);
  });

  it('throws on invalid token creation - too short', () => {
    const invalidBytes = new Uint8Array(31).fill(0x42);
    assert.throws(
      () => new BackupForwardSecrecyToken(invalidBytes),
      'Length of array supplied was 31 expected 32'
    );
  });

  it('throws on invalid token creation - too long', () => {
    const invalidBytes = new Uint8Array(33).fill(0x42);
    assert.throws(
      () => new BackupForwardSecrecyToken(invalidBytes),
      'Length of array supplied was 33 expected 32'
    );
  });

  it('supports round-trip serialization', () => {
    // Use different hardcoded pattern to ensure we're not just getting lucky
    const originalBytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      originalBytes[i] = i % 256;
    }

    const token = new BackupForwardSecrecyToken(originalBytes);

    // Serialize and deserialize
    const serialized = token.serialize();
    const reconstructedToken = new BackupForwardSecrecyToken(serialized);

    assert.exists(reconstructedToken);
    assert.equal(reconstructedToken.serialize().length, 32);

    const reconstructedBytes = reconstructedToken.serialize();
    expect(reconstructedBytes).to.deep.equal(originalBytes);
  });

  it('has correct SIZE constant', () => {
    assert.equal(BackupForwardSecrecyToken.SIZE, 32);
  });
});
